kvprober: enable kvprober metamorphically, deflake kvprober_integration_test.go

joshimhoff opened the pull request:
Fixes https://github.com/cockroachdb/cockroach/issues/106835. I ran the tests 100 times, and all runs were successful.

Side note: If https://github.com/cockroachdb/cockroach/issues/105807 is closed, kvprober will be enabled metamorphically in roachtests too. I could put up a more ad-hoc PR enabling kvprober from roachtests, if KV thinks that is desirable. Let me know.

**kvprober: metamorphically enable / configure kvprober**

This commit metamorphically enables & configures kvprober. Though kvprober
is off by default & not documented publicly, we run with kvprober enabled in
CC. So we should test CRDB with kvprober enabled. This commit is
inspired by the crdb_internal.probe_ranges corruption bug, tho it targets
kvprober proper, not crdb_internal.probe_ranges.

This commit also adjusts the kvprober default intervals to what they are set at
in CC. This is mostly done to improve the quality of tests with kvprober
enabled.

Release note: None.

**kvprober: slightly improve integration test by enabling quarantine probing**

Release note: None.

**kvprober: deflake kvprober_integration_test.go**

This commit deflakes TestProberDesReadsAndWrites. Multiple test cases asserted
that N probes would run, but in reality more than N may run, since kvprober
is both called in the test and run in the background as part of the test
server.

This commit also deflakes TestPlannerMakesPlansCoveringAllRanges. To ensure
the condition checked in the test is hit, it is necessary to create a
kvprober without ever calling Start.

Release note: None.

joshimhoff committed:
kvprober: metamorphically enable / configure kvprober

This commit metamorphically enables & configures kvprober. Though kvprober
is off by default & not documented publicly, we run with kvprober enabled in
CC. So we should test CRDB with kvprober enabled. This commit is
inspired by the crdb_internal.probe_ranges corruption bug, tho it targets
kvprober proper, not crdb_internal.probe_ranges.

This commit also adjusts the kvprober default intervals to what they are set at
in CC. This is mostly done to improve the quality of tests with kvprober
enabled.

Release note: None.

joshimhoff committed:
kvprober: slightly improve integration test by enabling quarantine probing

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106828)
<!-- Reviewable:end -->

joshimhoff committed:
kvprober: deflake kvprober_integration_test.go

This commit deflakes TestProberDesReadsAndWrites. Multiple test cases asserted
that N probes would run, but in reality more than N may run, since kvprober
is both called in the test and run in the background as part of the test
server.

This commit also deflakes TestPlannerMakesPlansCoveringAllRanges. To ensure
the condition checked in the test is hit, it is necessary to create a
kvprober without ever calling Start.

Release note: None.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-:-N_ZkCwW2kc21pYXwerx:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @joshimhoff)

___
*[`pkg/kv/kvprober/settings.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-N_ZjQC78rlhIWr2ppat:-N_ZjQC78rlhIWr2ppau:b8euc41) ([raw file](https://github.com/cockroachdb/cockroach/blob/99e343a5616d2033d6d44f4bb333609f77e5bb53/pkg/kv/kvprober/settings.go#L21)):*
> ```Go
> )
> 
> var enabled = util.ConstantWithMetamorphicTestBool("kv.prober.*.enabled", false)
> ```

nit: consider renaming to `defaultEnabled` so that this can't be confused with the final value.
___
*[`pkg/kv/kvprober/kvprober_integration_test.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-N_Zk6Ld0SB2kKxjp7kV:-N_Zk6Ld0SB2kKxjp7kW:b-3d0d4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/385621e96dc7698ded0868628f95b585dcf149a3/pkg/kv/kvprober/kvprober_integration_test.go#L177)):*
> ```Go
> 		require.Equal(t, p.Metrics().ReadProbeAttempts.Count(), p.Metrics().ReadProbeFailures.Count())
> 		// kvprober is running in background, so more than ten probes may be run.
> 		require.Greater(t, p.Metrics().ReadProbeFailures.Count(), int64(9))
> ```

nit: any reason not to use `GreaterOrEqual`?


<!-- Sent from Reviewable.io -->

joshimhoff reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-:-N_ZqwzsCrN-Mg6HCh0F:b-80vssw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvprober/settings.go` line 21 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-N_ZjQC78rlhIWr2ppat:-N_ZqQg93_PkaQ2e6Avq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99e343a5616d2033d6d44f4bb333609f77e5bb53/pkg/kv/kvprober/settings.go#L21)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider renaming to `defaultEnabled` so that this can't be confused with the final value.
</blockquote></details>

Done.
___
*[`pkg/kv/kvprober/kvprober_integration_test.go` line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/106828#-N_Zk6Ld0SB2kKxjp7kV:-N_ZqRJN4e9de9KzSeVO:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/385621e96dc7698ded0868628f95b585dcf149a3/pkg/kv/kvprober/kvprober_integration_test.go#L177)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: any reason not to use `GreaterOrEqual`?
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

joshimhoff commented:
bors r+

joshimhoff commented:
bors r+
